herbert_core/utilities/maths/noisify.m

```diff
@@ -17,40 +17,69 @@
 %           Add noise with Poisson distribution, where the mean value at
 %           a point is the value y.
 %
-%   If no input errors, e, just set e=[]
+%   >> [yout, eout] = noisify(y,e,[factor,]'maximum_value',maxval)
+%           Add noise with Gaussian distribution, calculating the standard
+%           deviation by an externally provided maximum y value.
+%           The max value is preceded by a keyword string 'maximum_value'.
+%           Typically this value will be the overall maximum value if the
+%           data is processed page by page and the maximum must be
+%           extracted before this processing.
+%           Omitting factor will use the default value 0.1.
+%
+%   If no input errors, e, just set e=[]. Note that eout will be created
+%   regardless of whether e is present, so the first overload output is
+%   probably incomplete.
+
+% deal with optional arguments
+is_poisson= false;
+fac = 0.1;
+if nargin>=2
+    p = inputParser;
+    addRequired(p,'y');   % y compulsory
+    addRequired(p,'e');   % e compulsory if any of remaining optional/parameter arguments present
+    numeric_or_string = @(x) isnumeric(x) || ischar(x);
+    addOptional(p,'dist_or_factor',fac, numeric_or_string);  % fac (numeric) or distribution ('poisson') optional, default to fac=0.1
+    addParameter(p,'maximum_value',-999,@isnumeric);  % ymax, as parameter 'maximum_value'. Default to highly visible negative value.
```

Comment:
Could the default here be `max(y)` rather than -999?

Reply:
If we can avoid introducing magic numbers it would be preferable.

Reply:
If max(y) is done here then it will be wasted cpu time for the poisson case (also it is max(abs(y(:)))). I can name the parameter - -999 was a distinctive visual value, its only effect here is to be negative and not difficult to read like -1. What I can do is name the constant in addition so it is not so much a magic number.

Reply:
Good point - I hadn't spotted we weren't using it in the Poisson case. Would `-inf` be an even more obvious negative value?

Reply:
That's a useful alternative. My measure of obvious/distinctive is "hits the eye of the beholder". Guessing that -999 is more "visual" than the string -inf - this is what you will see in the debugger. I was holding off using -666. But I would combine this with a named variable, I have coded USE_LOCAL_MAX to be visually distinctive in the code as distinct from the debugger.